src/station.cpp

```diff
@@ -305,81 +342,161 @@ Rect Station::GetCatchmentRect() const
+ * @param ind Industry
+ * @param st Station
  */
-static bool FindIndustryToDeliver(TileIndex ind_tile, void *user_data)
+static void AddIndustryToDeliver(Industry *ind, Station *st)
 {
-	/* Only process industry tiles */
-	if (!IsTileType(ind_tile, MP_INDUSTRY)) return false;
-
-	RectAndIndustryVector *riv = (RectAndIndustryVector *)user_data;
-	Industry *ind = Industry::GetByTile(ind_tile);
-
 	/* Don't check further if this industry is already in the list */
-	if (riv->industries_near->Contains(ind)) return false;
-
-	/* Only process tiles in the station acceptance rectangle */
-	int x = TileX(ind_tile);
-	int y = TileY(ind_tile);
-	if (x < riv->rect.left || x > riv->rect.right || y < riv->rect.top || y > riv->rect.bottom) return false;
+	if (st->industries_near.Contains(ind)) return;
 
 	/* Include only industries that can accept cargo */
 	uint cargo_index;
 	for (cargo_index = 0; cargo_index < lengthof(ind->accepts_cargo); cargo_index++) {
 		if (ind->accepts_cargo[cargo_index] != CT_INVALID) break;
 	}
-	if (cargo_index >= lengthof(ind->accepts_cargo)) return false;
+	if (cargo_index >= lengthof(ind->accepts_cargo)) return;
+
+	*st->industries_near.Append() = ind;
+}
 
-	*riv->industries_near->Append() = ind;
+/**
+ * Remove this station from the nearby stations lists of all towns and industries.
+ */
+void Station::RemoveFromAllNearbyLists()
+{
+	Town *t;
+	FOR_ALL_TOWNS(t) t->stations_near.erase(this);
+	Industry *i;
+	FOR_ALL_INDUSTRIES(i) i->stations_near.erase(this);
+}
 
+/**
+ * Test if the given town ID is covered by our catchment area.
+ * This is used when removing a house tile to determine if it was the last house tile
+ * within our catchment.
+ * @param t TownID to test.
+ * @return true if at least one house tile of TownID is covered.
+ */
+bool Station::CatchmentCoversTown(TownID t)
+{
+	for (auto it = this->catchment_tiles.begin(); it != this->catchment_tiles.end(); ++it) {
+		TileIndex tile = *it;
+		if (IsTileType(tile, MP_HOUSE) && GetTownIndex(tile) == t) return true;
+	}
 	return false;
 }
 
+/**
+ * Recompute tiles covered in our catchment area.
+ * This will additionally recompute nearby towns and industries.
+ */
+void Station::RecomputeCatchment()
+{
+	this->catchment_tiles.clear();
+
+	if (this->rect.IsEmpty()) return;
+
+	/* Loop finding all station tiles */
+	TileArea ta(TileXY(this->rect.left, this->rect.top), TileXY(this->rect.right, this->rect.bottom));
+	TILE_AREA_LOOP(tile, ta) {
+		if (!IsTileType(tile, MP_STATION) || GetStationIndex(tile) != this->index) continue;
+
+		uint r = GetTileCatchmentRadius(tile, this);
+		if (r == CA_NONE) continue;
+
+		/* This tile sub-loop doesn't need to test any tiles, they are simply added to the catchment set. */
+		TileArea ta2(TileXY(max<int>(TileX(tile) - r, 0), max<int>(TileY(tile) - r, 0)), TileXY(min<int>(TileX(tile) + r, MapMaxX()), min<int>(TileY(tile) + r, MapMaxY())));
+		TILE_AREA_LOOP(tile2, ta2) this->catchment_tiles.insert(tile2);
```

Comment:
Perhaps this could be improved by somehow finding contiguous areas with the same catchment radius, to avoid adding the same tiles multiple times.

Reply:
If RecomputeCatchment is only called each time a tile of the station is modified, maybe it isn't worth it by now.

Maybe creating a new function to extend a TileArea by a radius of r would be good here?

Reply:
I think with the new Bitmap type this is no longer a concern.